config/configload/load.go

```diff
@@ -48,218 +51,255 @@ func LoadBytes(src []byte, filename string) (*config.CouperFile, error) {
 	return LoadConfig(hclBody, src)
 }
 
-func LoadConfig(body hcl.Body, src []byte) (*config.CouperFile, error) {
+func LoadConfig(body hcl.Body, src []byte) (*config.Couper, error) {
 	defaults := config.DefaultSettings
-	file := &config.CouperFile{
+	couperConfig := &config.Couper{
 		Bytes:       src,
 		Context:     eval.NewENVContext(src),
 		Definitions: &config.Definitions{},
 		Settings:    &defaults,
 	}
 
-	fileSchema, _ := gohcl.ImpliedBodySchema(file)
-	content, diags := body.Content(fileSchema)
+	schema, _ := gohcl.ImpliedBodySchema(couperConfig)
+	content, diags := body.Content(schema)
 	if content == nil {
 		return nil, fmt.Errorf("invalid configuration: %w", diags)
 	}
 
-	// reading possible reference definitions first. Those are the base for refinement merges during server block read out.
-	var backends Backends
-	for _, outerBlock := range content.Blocks {
-		switch outerBlock.Type {
-		case definitions:
-			backendContent, leftOver, diags := outerBlock.Body.PartialContent(backendBlockSchema)
-
-			if diags.HasErrors() {
-				return nil, diags
-			}
-
-			if backendContent != nil {
-				for _, be := range backendContent.Blocks {
-					if len(be.Labels) == 0 {
-						return nil, hcl.Diagnostics{&hcl.Diagnostic{
-							Severity: hcl.DiagError,
-							Summary:  "Missing backend name",
-							Subject:  &be.DefRange,
-						}}
-					}
-					name := be.Labels[0]
-					ref, _ := backends.WithName(name)
-					if ref != nil {
-						return nil, hcl.Diagnostics{&hcl.Diagnostic{
-							Severity: hcl.DiagError,
-							Summary:  fmt.Sprintf("duplicate backend name: %q", name),
-							Subject:  &be.LabelRanges[0],
-						}}
-					}
-					backends = append(backends, NewBackend(name, be.Body))
-				}
-			}
+	// Read possible reference definitions first. Those are the
+	// base for refinement merges during server block read out.
+	backends, err := loadDefinitions(couperConfig, content.Blocks)
```

Comment:
Splitting this loop by type into two methods with their own switch case for one type does not help here and loses the context. For readability during load, this should stay here as it is.